hermes/bin/src/runtime_extensions/hermes/ipfs/state/api.rs

```diff
@@ -0,0 +1,128 @@
+            is_valid
+        },
+    }
+}
+
+/// Get File from Ipfs
+pub(crate) fn hermes_ipfs_get_file(
+    app_name: &HermesAppName, path: &IpfsPath,
+) -> Result<IpfsFile, Errno> {
+    tracing::debug!(app_name = %app_name, path = %path, "get IPFS file");
+    let content = HERMES_IPFS_STATE.file_get(path)?;
+    tracing::debug!(app_name = %app_name, path = %path, "got IPFS file");
+    Ok(content)
+}
+
+/// Pin IPFS File
+pub(crate) fn hermes_ipfs_pin_file(
+    app_name: &HermesAppName, path: IpfsPath,
+) -> Result<bool, Errno> {
+    tracing::debug!(app_name = %app_name, path = %path, "pin IPFS file");
+    let status = HERMES_IPFS_STATE.file_pin(&path)?;
+    tracing::debug!(app_name = %app_name, path = %path, "pinned IPFS file");
+    HERMES_IPFS_STATE.apps.pinned_file(app_name.clone(), path);
+    Ok(status)
+}
+
+/// Get DHT Value
+pub(crate) fn hermes_ipfs_get_dht_value(
+    app_name: &HermesAppName, key: DhtKey,
+) -> Result<DhtValue, Errno> {
+    let key_str = format!("{key:x?}");
+    tracing::debug!(app_name = %app_name, dht_key = %key_str, "get DHT value");
+    let value = HERMES_IPFS_STATE.dht_get(key)?;
+    tracing::debug!(app_name = %app_name, dht_key = %key_str, "got DHT value");
+    Ok(value)
+}
+
+/// Put DHT Value
+pub(crate) fn hermes_ipfs_put_dht_value(
+    app_name: &HermesAppName, key: DhtKey, value: DhtValue,
+) -> Result<bool, Errno> {
+    let key_str = format!("{key:x?}");
+    tracing::debug!(app_name = %app_name, dht_key = %key_str, "putting DHT value");
+    let status = HERMES_IPFS_STATE.dht_put(key.clone(), value)?;
+    tracing::debug!(app_name = %app_name, dht_key = %key_str, "have put DHT value");
+    HERMES_IPFS_STATE.apps.added_dht_key(app_name.clone(), key);
+    Ok(status)
+}
+
+/// Subscribe to a topic
+pub(crate) fn hermes_ipfs_subscribe(
+    app_name: &HermesAppName, topic: PubsubTopic,
+) -> Result<bool, Errno> {
+    tracing::debug!(app_name = %app_name, pubsub_topic = %topic, "subscribing to PubSub topic");
+    if HERMES_IPFS_STATE.apps.topic_subscriptions_contains(&topic) {
+        tracing::debug!(app_name = %app_name, pubsub_topic = %topic, "topic subscription stream already exists");
+    } else {
+        let handle = HERMES_IPFS_STATE.pubsub_subscribe(&topic)?;
+        HERMES_IPFS_STATE
+            .apps
+            .added_topic_stream(topic.clone(), handle);
+        tracing::debug!(app_name = %app_name, pubsub_topic = %topic, "added subscription topic stream");
+    }
+    HERMES_IPFS_STATE
+        .apps
+        .added_app_topic_subscription(app_name.clone(), topic);
+    Ok(true)
+}
+
+/// Publish message to a topic
+pub(crate) fn hermes_ipfs_publish(
+    _app_name: &HermesAppName, topic: &PubsubTopic, message: MessageData,
+) -> Result<MessageId, Errno> {
+    let message_id = HERMES_IPFS_STATE
+        .pubsub_publish(topic.to_string(), message)
+        .map(|m| m.0)?;
+    Ok(message_id.0)
```

Comment:
Can it be something like this? So don't need to wrap the value with `Ok`.

```suggestion
    HERMES_IPFS_STATE
        .pubsub_publish(topic.to_string(), message)
        .map(|m| m.0.0)
```